src/fluree/db/query/optimize.cljc

```diff
@@ -65,3 +67,221 @@
+    mch))
+
+(declare compile-filter-codes)
+
+(defn compile-pattern-filters
+  "Recursively compile filter codes in a pattern."
+  [pattern context]
+  (let [pattern-type (where/pattern-type pattern)]
+    (case (where/pattern-type pattern)
+
+      :tuple
+      ;; Tuple patterns are vectors of match objects
+      (mapv #(compile-inline-filters % context) pattern)
+
+      (:class :id)
+      ;; Class and ID patterns have a single match object as data
+      (let [mch (-> pattern
+                    where/pattern-data
+                    (compile-inline-filters context))]
+        (where/->pattern pattern-type mch))
+
+      :union
+      ;; Union patterns contain a vector of where clauses
+      (let [clauses (->> (where/pattern-data pattern)
+                         (mapv (partial compile-filter-codes context)))]
+        (where/->pattern pattern-type clauses))
+
+      (:optional :exists :not-exists :minus)
+      ;; Optional, exists, not-exists, and minus patterns contain a single where
+      ;; clause
+      (let [where-clause (->> (where/pattern-data pattern)
+                              (compile-filter-codes context))]
+        (where/->pattern pattern-type where-clause))
+
+      :graph
+      ;; Graph patterns contain [graph* where-clause]
+      (let [[graph where-clause] (where/pattern-data pattern)
+            where-clause* (compile-filter-codes context where-clause)]
+        (where/->pattern pattern-type [graph where-clause*]))
+
+      ;; All other pattern types pass through unchanged
+      pattern)))
+
+(defn compile-filter-codes
+  "Walk through where clause and compile all filter codes."
+  [context where-clause]
+  (if (seq where-clause)
+    (mapv #(compile-pattern-filters % context) where-clause)
+    where-clause))
+
+(defn extract-var-filters
+  "Extract a map of variable -> filter function for single-variable filters."
+  [where-clause]
+  (reduce (fn [acc pattern]
+            (if-let [variable (get-filtered-variable pattern)]
+              (assoc acc variable (where/pattern-data pattern))
+              acc))
+          {} where-clause))
+
+(defn find-filtered-vars
+  "Find variables that the tuple binds and have filters."
+  [tuple var-filters]
+  (filter (partial tuple-binds-var? tuple)
+          (keys var-filters)))
+
+(defn attach-filters
+  "Attach filter codes to a tuple for the given filtered variables."
+  [tuple filtered-vars var-filters]
+  (reduce (fn [tuple* variable]
+            (let [codes (-> var-filters (get variable) get-filter-codes)]
+              (with-var-filter tuple* variable codes)))
+          tuple filtered-vars))
+
+(defn process-binding-pattern
+  "Process a binding pattern (tuple/class/id) and attach inline filters.
+  Returns [processed-pattern filtered-vars]."
+  [pattern pattern-type var-filters]
+  (let [tuple (if (= :tuple pattern-type)
+                pattern
+                (where/pattern-data pattern))
```

Comment:
Should we be processing non-tuple patterns at all? My thought is that we should be looking for a filter pattern that comes after one or more tuple patterns, recursively.

Reply:
we also need to consider `:class` and `:id` patterns too. This function is only called on those pattern types as well as those of type `:tuple`